File: ARMController/app/src/main/java/com/pawelcembaluk/armcontroller/bluetooth/BufferedData.java
package com.pawelcembaluk.armcontroller.bluetooth;

import java.util.Arrays;

public class BufferedData {

    private final StringBuilder buffer = new StringBuilder();

    public void append(String string) {
        buffer.append(string);
    }

    public String[] flush() {
        String bufferedString = buffer.toString();
        if (endsWithNewLine(bufferedString))
            return flushAllData();
        else
            return flushOnlyFullLines();
    }

    private boolean endsWithNewLine(String dataString) {
        return dataString.matches("(?s).*\\R$");
    }

    private String[] flushAllData() {
        String[] dataStrings = splitLines(buffer.toString());
        cleanBuffer();
        return dataStrings;
    }

    private String[] splitLines(String string) {
        return string.split("\\R");
    }

    private void cleanBuffer() {
        buffer.setLength(0);
    }

    private String[] flushOnlyFullLines() {
        String[] dataStrings = splitLines(buffer.toString());
        if (dataStrings.length < 2)
            return new String[0]; //No lines or only one line that doesn't end with \R.
        cleanBuffer();
        String lastElement = dataStrings[dataStrings.length - 1];
        buffer.append(lastElement);
        return Arrays.copyOfRange(dataStrings, 0, dataStrings.length - 1);
    }
}

File: ARM/connection/BufferedData.py
import re


class BufferedData:
    def __init__(self):
        self.buffer = ""

    def append(self, string):
        self.buffer += string

    def flush(self):
        return self.flushAllData() if self.endsWithNewLine(self.buffer) else self.flushOnlyFullLines()

    def flushAllData(self):
        dataStrings = self.buffer.splitlines()
        self.buffer = ""
        return dataStrings

    def endsWithNewLine(self, string):
        return re.fullmatch(".*[\r\n]+", string, re.S)

    def flushOnlyFullLines(self):
        dataStrings = self.buffer.splitlines()
        self.buffer = dataStrings[-1]  # last string
        return dataStrings[:-1]  # everything except last string
